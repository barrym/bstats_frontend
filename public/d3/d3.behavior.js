(function(){d3.behavior = {};
// TODO unbind zoom behavior?
// TODO unbind listener?
d3.behavior.zoom = function() {
  var x = 0,
      y = 0,
      z = 0,
      panning,
      zooming,
      locations = {}, // identifier -> location
      last = 0,
      target,
      targuments,
      event = d3.dispatch("zoom");

  // mousewheel events are totally broken!
  // https://bugs.webkit.org/show_bug.cgi?id=40441
  // not only that, but Chrome and Safari differ in re. to acceleration!
  var div = d3.select("body").append("div")
      .style("visibility", "hidden")
      .style("top", 0)
      .style("height", 0)
      .style("width", 0)
      .style("overflow-y", "scroll")
    .append("div")
      .style("height", "2000px")
    .node().parentNode;

  function zoom() {
    this
        .on("mousedown.zoom", mousedown)
        .on("mousewheel.zoom", mousewheel)
        .on("DOMMouseScroll.zoom", dblclick)
        .on("dblclick.zoom", dblclick)
        .on("touchstart.zoom", touchstart);

    d3.select(window)
        .on("mousemove.zoom", mousemove)
        .on("mouseup.zoom", mouseup)
        .on("touchmove.zoom", touchmove)
        .on("touchend.zoom", touchup);
  }

  function location(point) {
    return [point[0] - x, point[1] - y, z];
  }

  // snapshot the target, data and index for subsequent dispatch
  function start() {
    target = this;
    targuments = arguments;
  }

  // Note: Since we don't rotate, it's possible for the touches to become
  // slightly detached from their original positions. Thus, we recompute the
  // touch points on touchend as well as touchstart!
  function touchup() {
    var touches = d3.svg.touches(target),
        i = -1,
        n = touches.length,
        touch;
    while (++i < n) locations[(touch = touches[i]).identifier] = location(touch);
    return touches;
  }

  function touchstart() {
    start.apply(this, arguments);
    var touches = touchup(),
        touch,
        now = Date.now();

    // doubletap detection
    if ((touches.length === 1) && (now - last < 300)) {
      var touch = touches[0];
      zoomto(1 + Math.floor(z), touch, locations[touch.identifier]);
    }
    last = now;
  }

  function touchmove() {
    var touches = d3.svg.touches(target);
    switch (touches.length) {

      // single-touch pan
      case 1: {
        var touch = touches[0];
        zoomto(z, touch, locations[touch.identifier]);
        break;
      }

      // double-touch pan + zoom
      case 2: {
        var p0 = touches[0],
            p1 = touches[1],
            p2 = [(p0[0] + p1[0]) / 2, (p0[1] + p1[1]) / 2],
            l0 = locations[p0.identifier],
            l1 = locations[p1.identifier],
            l2 = [(l0[0] + l1[0]) / 2, (l0[1] + l1[1]) / 2, l0[2]];
        zoomto(Math.log(d3.event.scale) / Math.LN2 + l0[2], p2, l2);
        break;
      }
    }
  }

  function mousedown() {
    start.apply(this, arguments);
    panning = location(d3.svg.mouse(target));
    d3.event.preventDefault();
    window.focus();
  }

  function mousemove() {
    zooming = null;
    if (panning) zoomto(z, d3.svg.mouse(target), panning);
  }

  function mouseup() {
    if (panning) {
      mousemove();
      panning = null;
    }
  }

  function mousewheel() {
    start.apply(this, arguments);

    // store starting mouse location
    if (!zooming) zooming = location(d3.svg.mouse(target));

    // detect the pixels that would be scrolled by this wheel event
    var e = d3.event, delta;
    try {
      div.scrollTop = 1000;
      div.dispatchEvent(e);
      delta = 1000 - div.scrollTop;
    } catch (error) {
      delta = e.wheelDelta || -e.detail;
    }

    zoomto(delta * .005 + z, d3.svg.mouse(target), zooming);
  }

  function dblclick() {
    start.apply(this, arguments);
    var mouse = d3.svg.mouse(target);
    zoomto(d3.event.shiftKey ? Math.ceil(z - 1) : Math.floor(z + 1), mouse, location(mouse));
  }

  function zoomto(z1, x0, x1) {
    var k = Math.pow(2, (z = z1) - x1[2]);
    x = x0[0] - k * x1[0];
    y = x0[1] - k * x1[1];
    dispatch();
  }

  function dispatch() {
    var o = d3.event, // Events can be reentrant (e.g., focus).
        k = Math.pow(2, z);

    d3.event = {
      scale: k,
      translate: [x, y],
      transform: function(sx, sy) {
        if (sx) transform(sx, x);
        if (sy) transform(sy, y);
      }
    };

    function transform(scale, o) {
      var domain = scale.__domain || (scale.__domain = scale.domain()),
          range = scale.range().map(function(v) { return (v - o) / k; });
      scale.domain(domain).domain(range.map(scale.invert));
    }

    try {
      event.zoom.dispatch.apply(target, targuments);
    } finally {
      d3.event = o;
    }

    o.preventDefault();
  }

  zoom.on = function(type, listener) {
    event[type].add(listener);
    return zoom;
  };

  return zoom;
};
})();
