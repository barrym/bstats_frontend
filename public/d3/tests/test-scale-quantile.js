require("./../lib/env-js/envjs/node");
require("./../d3");

var x = d3.scale.quantile()
    .domain([3, 6, 7, 8, 8, 10, 13, 15, 16, 20])
    .range([0, 1, 2, 3]);

console.log("domain(3, 6, 7, 8, 8, 10, 13, 15, 16, 20).range(0, 1, 2, 3):");
console.log("  quantiles -> " + x.quantiles());
console.log("          3 -> " + x(3));
console.log("          6 -> " + x(6));
console.log("        6.9 -> " + x(6.9));
console.log("          7 -> " + x(7));
console.log("        7.1 -> " + x(7.1));
console.log("          8 -> " + x(8));
console.log("        8.9 -> " + x(8.9));
console.log("          9 -> " + x(9));
console.log("        9.1 -> " + x(9.1));
console.log("         10 -> " + x(10));
console.log("         13 -> " + x(13));
console.log("       14.9 -> " + x(14.9));
console.log("         15 -> " + x(15));
console.log("       15.1 -> " + x(15.1));
console.log("         16 -> " + x(16));
console.log("         20 -> " + x(20));
console.log("");

var x = d3.scale.quantile()
    .domain([3, 6, 7, 8, 8, 9, 10, 13, 15, 16, 20])
    .range([0, 1, 2, 3]);

console.log("domain(3, 6, 7, 8, 8, 9, 10, 13, 15, 16, 20).range(0, 1, 2, 3):");
console.log("  quantiles -> " + x.quantiles());
console.log("          3 -> " + x(3));
console.log("          6 -> " + x(6));
console.log("        6.9 -> " + x(6.9));
console.log("          7 -> " + x(7));
console.log("        7.1 -> " + x(7.1));
console.log("          8 -> " + x(8));
console.log("        8.9 -> " + x(8.9));
console.log("          9 -> " + x(9));
console.log("        9.1 -> " + x(9.1));
console.log("         10 -> " + x(10));
console.log("         13 -> " + x(13));
console.log("       14.9 -> " + x(14.9));
console.log("         15 -> " + x(15));
console.log("       15.1 -> " + x(15.1));
console.log("         16 -> " + x(16));
console.log("         20 -> " + x(20));
console.log("");

var x = d3.scale.quantile()
    .domain([0, 1])
    .range([0, 1]);

console.log("domain(0, 1).range(0, 1):");
console.log("  quantiles -> " + x.quantiles());
console.log("        -.5 -> " + x(-.5));
console.log("          0 -> " + x(0));
console.log("        .49 -> " + x(.49));
console.log("        .51 -> " + x(.51));
console.log("          1 -> " + x(1));
console.log("        1.5 -> " + x(1.5));
console.log("");

var x = d3.scale.quantile()
    .domain([1, 2, 3, 4])
    .range(["a", "b"]);

console.log("domain(1, 2, 3, 4).range(a, b):");
console.log("  quantiles -> " + x.quantiles());
console.log("          0 -> " + x(0));
console.log("          1 -> " + x(1));
console.log("          2 -> " + x(2));
console.log("       2.49 -> " + x(2.49));
console.log("       2.51 -> " + x(2.51));
console.log("          3 -> " + x(3));
console.log("          4 -> " + x(4));
console.log("          5 -> " + x(5));
console.log("");

var x = d3.scale.quantile()
    .domain([1, 2, 3, 4])
    .range(["a", "b", "c"]);

console.log("domain(1, 2, 3, 4).range(a, b, c):");
console.log("  quantiles -> " + x.quantiles());
console.log("          0 -> " + x(0));
console.log("          1 -> " + x(1));
console.log("        1.9 -> " + x(1.9));
console.log("          2 -> " + x(2));
console.log("        2.1 -> " + x(2.1));
console.log("        2.9 -> " + x(2.9));
console.log("          3 -> " + x(3));
console.log("        3.1 -> " + x(3.1));
console.log("          4 -> " + x(4));
console.log("          5 -> " + x(5));
console.log("");

var x = d3.scale.quantile()
    .domain([1, 1, 2, 2, 3, 3, 4, 4])
    .range(["a", "b", "c", "d"]);

console.log("domain(1, 1, 2, 2, 3, 3, 4, 4).range(a, b, c, d):");
console.log("  quantiles -> " + x.quantiles());
console.log("          0 -> " + x(0));
console.log("          1 -> " + x(1));
console.log("       1.49 -> " + x(1.49));
console.log("       1.51 -> " + x(1.51));
console.log("          2 -> " + x(2));
console.log("       2.49 -> " + x(2.49));
console.log("       2.51 -> " + x(2.51));
console.log("          3 -> " + x(3));
console.log("       3.49 -> " + x(3.49));
console.log("       3.51 -> " + x(3.51));
console.log("          4 -> " + x(4));
console.log("          5 -> " + x(5));
console.log("");

var x = d3.scale.quantile()
    .domain([1, 2, 3, 4, 5, 6, 7, 8])
    .range(["a", "b", "c", "d"]);

console.log("domain(1, 2, 3, 4, 5, 6, 7, 8).range(a, b, c, d):");
console.log("  quantiles -> " + x.quantiles());
console.log("          0 -> " + x(0));
console.log("          1 -> " + x(1));
console.log("          2 -> " + x(2));
console.log("       2.49 -> " + x(2.49));
console.log("       2.51 -> " + x(2.51));
console.log("          3 -> " + x(3));
console.log("          4 -> " + x(4));
console.log("       4.49 -> " + x(4.49));
console.log("       4.51 -> " + x(4.51));
console.log("          5 -> " + x(5));
console.log("          6 -> " + x(6));
console.log("       6.49 -> " + x(6.49));
console.log("       6.51 -> " + x(6.51));
console.log("          7 -> " + x(7));
console.log("          8 -> " + x(8));
console.log("          9 -> " + x(9));
console.log("");
