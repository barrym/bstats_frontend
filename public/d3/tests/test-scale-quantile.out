domain(3, 6, 7, 8, 8, 10, 13, 15, 16, 20).range(0, 1, 2, 3):
  quantiles -> 7.25,9,14.5
          3 -> 0
          6 -> 0
        6.9 -> 0
          7 -> 0
        7.1 -> 0
          8 -> 1
        8.9 -> 1
          9 -> 2
        9.1 -> 2
         10 -> 2
         13 -> 2
       14.9 -> 3
         15 -> 3
       15.1 -> 3
         16 -> 3
         20 -> 3

domain(3, 6, 7, 8, 8, 9, 10, 13, 15, 16, 20).range(0, 1, 2, 3):
  quantiles -> 7.5,9,14
          3 -> 0
          6 -> 0
        6.9 -> 0
          7 -> 0
        7.1 -> 0
          8 -> 1
        8.9 -> 1
          9 -> 2
        9.1 -> 2
         10 -> 2
         13 -> 2
       14.9 -> 3
         15 -> 3
       15.1 -> 3
         16 -> 3
         20 -> 3

domain(0, 1).range(0, 1):
  quantiles -> 0.5
        -.5 -> 0
          0 -> 0
        .49 -> 0
        .51 -> 1
          1 -> 1
        1.5 -> 1

domain(1, 2, 3, 4).range(a, b):
  quantiles -> 2.5
          0 -> a
          1 -> a
          2 -> a
       2.49 -> a
       2.51 -> b
          3 -> b
          4 -> b
          5 -> b

domain(1, 2, 3, 4).range(a, b, c):
  quantiles -> 2,3
          0 -> a
          1 -> a
        1.9 -> a
          2 -> b
        2.1 -> b
        2.9 -> b
          3 -> c
        3.1 -> c
          4 -> c
          5 -> c

domain(1, 1, 2, 2, 3, 3, 4, 4).range(a, b, c, d):
  quantiles -> 1.75,2.5,3.25
          0 -> a
          1 -> a
       1.49 -> a
       1.51 -> a
          2 -> b
       2.49 -> b
       2.51 -> c
          3 -> c
       3.49 -> d
       3.51 -> d
          4 -> d
          5 -> d

domain(1, 2, 3, 4, 5, 6, 7, 8).range(a, b, c, d):
  quantiles -> 2.75,4.5,6.25
          0 -> a
          1 -> a
          2 -> a
       2.49 -> a
       2.51 -> a
          3 -> b
          4 -> b
       4.49 -> b
       4.51 -> c
          5 -> c
          6 -> c
       6.49 -> d
       6.51 -> d
          7 -> d
          8 -> d
          9 -> d

